eng/common/scripts/git-branch-push.ps1

```diff
@@ -63,8 +63,10 @@ else
   }
 }
 # Checkout to $PRBranch, create new one if not exists.
-git show-ref --verify --quiet refs/heads/$PRBranchName
+git ls-remote --exit-code --heads $RemoteName $PRBranchName
 if ($LASTEXITCODE -eq 0) {
+  Write-Host "git fetch $RemoteName $PRBranchName"
+  git fetch $RemoteName $PRBranchName
   Write-Host "git checkout $PRBranchName."
   git checkout $PRBranchName 
```

Comment:
This checkout might fail now if the local changes aren't up to date with what you just fetched, which will be a lot of our normal cases. We are going to need to figure out a way to get this error to fall into our retry loop.

Reply:
I don't think this is a problem because even if we fetch the remote, until we merge/rebase the remote ref into our local branch state, `git checkout` will work fine. When there are conflicts, the retry loop lower down in the script will find and address them.

e.g. if I create a local state that is both behind the tip of a remote branch and has extra commits, I can still run fetch+checkout against that branch, e.g.

```
⇉ ⇉ ⇉ git checkout master
M       eng/README.md
Already on 'master'
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
```

Reply:
I tested scenarios where my local state had different and missing changes from upstream, and the rebase handling was able to reconcile them by rebasing the changes together (though perhaps we should walk through it to make sure it reconciled them in ways that are intended).

Reply:
The scenario I'm thinking about is one where there are new changes that you just pulled down with the fetch   that intersect with the changes you have currently in the working directory. In such a case the checkout would fail and we would exit the script and never hit the retry block. Until this point we never had to deal with that case because all the changes were already available locally and we generally were already on the branch we were going to checkout and commit to (granted maybe accidently in some cases). I do know we definitely hit the retry block in normal workflows when we release multiple packages at once so I want to make sure we aren't now hitting that case early and failing here where as before we would have retried. 

I wonder if we should simply update the fetch in the retry block and call this fix good.

Reply:
My understanding is the fetch will pull the changes into the `<remote>/<branch>` ref, and NOT merge them into the local ref `<branch>`. It only tries to reconcile the two when you get to the rebase stage of the script.

If I push a commit to a remote branch `benbp/foo` adding `foo` to the top of the file, then set my branch state to master without that commit, then make a similar change to the same line of the same file adding `bar`, the script has no problem reconciling the two:

```
⇉ ⇉ ⇉ git diff | tee
diff --git a/eng/README.md b/eng/README.md
index e80fb87c..10158f3b 100644
--- a/eng/README.md
+++ b/eng/README.md
@@ -1,3 +1,4 @@
+bar
 # Engineering Tools

 All the tools/utilities used in Microsoft Azure Java SDK's build config are defined here.

⇉ ⇉ ⇉ git fetch benbp benbp/foo
From github.com:benbp/azure-sdk-for-java
 * branch              benbp/foo  -> FETCH_HEAD

⇉ ⇉ ⇉ git status
On branch benbp/foo
Your branch is behind 'benbp/benbp/foo' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)

You are in a sparse checkout with 1% of tracked files present.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   eng/README.md

no changes added to commit (use "git add" and/or "git commit -a")

⇉ ⇉ ⇉ git diff HEAD..benbp/benbp/foo | tee
diff --git a/eng/README.md b/eng/README.md
index e80fb87c..26a043d7 100644
--- a/eng/README.md
+++ b/eng/README.md
@@ -1,3 +1,4 @@
+foo
 # Engineering Tools

 All the tools/utilities used in Microsoft Azure Java SDK's build config are defined here.
```

Running the script with local conflicting changes on an out of date branch:

```
$ /home/ben/testing/refs/git-branch-push.ps1 -PRBranchName benbp/foo -CommitMsg "foo rebased" -GitUrl git@github.com:benbp/azure-sdk-for-java
git remote add azure-sdk-fork git@github.com:benbp/azure-sdk-for-java
3054ce64fd05031fca9d5a0f2c597b9668bee622        refs/heads/benbp/foo
git fetch azure-sdk-fork benbp/foo
From github.com:benbp/azure-sdk-for-java
 * branch              benbp/foo  -> FETCH_HEAD
 * [new branch]        benbp/foo  -> azure-sdk-fork/benbp/foo
git checkout benbp/foo.
M       eng/README.md
Already on 'benbp/foo'
Your branch is behind 'benbp/benbp/foo' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" commit  -am "foo rebased"
[benbp/foo 7a574a40] foo rebased
 1 file changed, 1 insertion(+)
git push azure-sdk-fork benbp/foo
To github.com:benbp/azure-sdk-for-java
 ! [rejected]          benbp/foo -> benbp/foo (non-fast-forward)
error: failed to push some refs to 'github.com:benbp/azure-sdk-for-java'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
Git push failed with LASTEXITCODE=1 Need to fetch and rebase: attempt number=1
git fetch azure-sdk-fork benbp/foo
From github.com:benbp/azure-sdk-for-java
 * branch              benbp/foo  -> FETCH_HEAD
git diff benbp/foo~ benbp/foo --output /tmp/tmpPFi3gw.tmp
git reset --hard azure-sdk-fork/benbp/foo
HEAD is now at 3054ce64 foo
git apply -C0 /tmp/tmpPFi3gw.tmp
git add -A
git -c user.name="azure-sdk" -c user.email="azuresdk@microsoft.com" commit -m "foo rebased"
[benbp/foo 123f6394] foo rebased
 1 file changed, 1 insertion(+)
git push azure-sdk-fork benbp/foo
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 8 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 366 bytes | 183.00 KiB/s, done.
Total 4 (delta 3), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To github.com:benbp/azure-sdk-for-java
   3054ce64..123f6394  benbp/foo -> benbp/foo

$ git log -n 2 | tee
commit 123f6394dc64e5c835986f7fb407e9fd2fa7b2e5
Author: azure-sdk <azuresdk@microsoft.com>
Date:   Tue Jun 1 19:53:32 2021 -0400

    foo rebased

commit 3054ce64fd05031fca9d5a0f2c597b9668bee622
Author: Ben Broderick Phillips <bebroder@microsoft.com>
Date:   Tue Jun 1 19:47:34 2021 -0400

    foo

$ git show | tee
commit 123f6394dc64e5c835986f7fb407e9fd2fa7b2e5
Author: azure-sdk <azuresdk@microsoft.com>
Date:   Tue Jun 1 19:53:32 2021 -0400
                                                                                                                                                                                                                                   foo rebased

diff --git a/eng/README.md b/eng/README.md
index 26a043d7..8cfe3fb5 100644
--- a/eng/README.md
+++ b/eng/README.md
@@ -1,4 +1,5 @@
 foo
+bar
 # Engineering Tools

 All the tools/utilities used in Microsoft Azure Java SDK's build config are defined here.
```

Reply:
But that said, the multiple fetches are redundant, so we can probably simplify this script as you say. Let me take a look.

Reply:
The conflict doesn't occur on fetch but instead on checkout. If the checkout is creating the branch (i.e. it existed remote but not locally) and then you fetch down a change that is in conflict with what is currently in the working directory or a staged change. Then the checkout will fail with an error such as:

```
> git checkout Test1
error: Your local changes to the following files would be overwritten by checkout:
        README.md
Please commit your changes or stash them before you switch branches.
Aborting
```

If we are in this state we will never hit the retry loop and we will fail at the checkout step before the retry. This is why I like the idea of just having the fetch in the retry loop.

Reply:
Updated now with @weshaggard. Looks like the issue was just doing a global fetch when in a sparse state, and others changes were unnecessary.